#12, #78: Extend command API

martin-fleck-at opened the pull request:
- Make client commands more extensible
-- Use String for type argument instead of enum
-- Add generic properties to provide custom information

- Make command codec injectable with client contributions
-- Clients can contribute translation for their commands based on type
-- Move previous implementation to custom EMF command contributions
-- Move shared code into dedicated utility class
-- Provide EMFCommandCodec for easier non-DI usability

- Use transactional editing domain and command stack
-- Perform model update on command stack (is now undoable)
-- Add EMF Transaction to target platform and maven build

- Extend Command model with CommandExecutionResult
-- Provide original command and affected EObject in result
-- Provide change recordings in result
-- Allow for custom type and details map in result

- Remove 'edit' command with EMF commands from Java Client API

- Add custom command to Coffee example
-- Command simply adds the given text to all tasks

- Simply discard dangling references on XMI serialization

Client adaptation in https://github.com/eclipse-emfcloud/emfcloud-modelserver-theia/pull/35

ndoschek reviewed (changes requested):
Thank you very much for this excellent contribution Martin!
Overall the functionality seems to work great, and I am a huge fan of the re-worked "command concept"! :+1: 
I am currently still in the process of testing, but meanwhile I found some superficial issues, which might be fixed rather quickly I assume.

martin-fleck-at commented on bundles/org.eclipse.emfcloud.modelserver.edit/src/org/eclipse/emfcloud/modelserver/edit/EMFCommandCodec.java:
```diff
@@ -0,0 +1,177 @@
+      return result;
+   }
+
+   @SuppressWarnings("checkstyle:CyclomaticComplexity")
+   public static Command invert(final Command command) {
+      if (command instanceof CompoundCommand) {
+         CompoundCommand undoCompoundCommand = (CompoundCommand) command;
+         CompoundCommand inverseCompoundCommand = new CompoundCommand();
+         for (Command next : undoCompoundCommand.getCommandList()) {
+            inverseCompoundCommand.append(invert(next));
+         }
+         return inverseCompoundCommand;
+      } else if (command instanceof AddCommand) {
+         AddCommand undoAddCommand = (AddCommand) command;
+         return RemoveCommand.create(undoAddCommand.getDomain(), undoAddCommand.getOwner(),
+            undoAddCommand.getFeature(), undoAddCommand.getResult());
+      } else if (command instanceof RemoveCommand) {
+         RemoveCommand undoRemoveCommand = (RemoveCommand) command;
+         return AddCommand.create(undoRemoveCommand.getDomain(), undoRemoveCommand.getOwner(),
+            undoRemoveCommand.getFeature(), undoRemoveCommand.getResult(), undoRemoveCommand.getIndices()[0]);
+      } else if (command instanceof SetCommand) {
+         SetCommand undoSetCommand = (SetCommand) command;
+         return SetCommand.create(undoSetCommand.getDomain(), undoSetCommand.getOwner(),
+            undoSetCommand.getFeature(), undoSetCommand.getOldValue(), undoSetCommand.getIndex());
+      }
+      return command;
+   }
```
We don't use it anymore, no. I just kept it for people who might still want to use it. But maybe it is better to remove it, you are right.

martin-fleck-at commented on bundles/org.eclipse.emfcloud.modelserver.emf/src/org/eclipse/emfcloud/modelserver/emf/common/DefaultModelController.java:
```diff
@@ -216,49 +215,33 @@ public void getValidationConstraints(final Context ctx, final String modeluri) {
+      Optional<CCommandExecutionResult> undoExecution = modelRepository.undo(modeluri);
+      if (undoExecution.isPresent()) {
+         success(ctx, "Successful undo.");
+         Map<String, JsonNode> encodings = getCommandEncodings(modeluri, undoExecution.get());
+         sessionController.commandExecuted(modeluri, encodings);
```
Oh yeah, sure, good idea!

martin-fleck-at commented on bundles/org.eclipse.emfcloud.modelserver.emf/src/org/eclipse/emfcloud/modelserver/emf/common/ModelRepository.java:
```diff
@@ -60,17 +61,13 @@
-   void addTemporaryCommandResource(String modeluri, Resource resource, CCommand command);
-
-   void removeTemporaryCommandResource(String modeluri, Resource resource);
+   Optional<CCommandExecutionResult> redo(String modeluri);
```
I cleaned those methods up, thank you!

martin-fleck-at commented on bundles/org.eclipse.emfcloud.modelserver.emf/src/org/eclipse/emfcloud/modelserver/emf/common/ModelServerEditingDomain.java:
```diff
@@ -56,11 +55,9 @@ public Command getUndoCommand() {
        */
       if (canUndo()) {
-         return createInverseCommand(commandStack.getUndoCommand());
+         return commandStack.getUndoCommand();
```
I modified the code ab it and cleaned it up, thank you!

martin-fleck-at commented:
@ndoschek Thank you very much for your review, I took another look at the things you mentioned and pushed a new follow-up commit that fixes several things. I also fixed the test cases, at least for me locally they run through :-)

ndoschek reviewed (changes requested):
Thank you again for updating the PR and especially for fixing the tests! :+1: 
So far, the functionality looks great to me and it is quite intuitive to adapt to the new concept!
But I noticed a few minor things:
- The `XmiToEObjectSubsubscriptionListener` needs to be updated as well as it still excpects a `CCommand` instead of the new result type.
- Please update the README
- It would be great if we provided a custom command example in Java somewhere, either in the README and/or in the `ExampleModelServerSubscriptionClient` - what do you think?

martin-fleck-at commented on bundles/org.eclipse.emfcloud.modelserver.emf/src/org/eclipse/emfcloud/modelserver/emf/common/DefaultSessionController.java:
```diff
@@ -124,13 +125,13 @@ public void modelUpdated(final String modeluri) {
 
    @Override
-   public void commandExecuted(final String modeluri, final Map<String, JsonNode> encodings) {
+   public void commandExecuted(final String modeluri, final CCommandExecutionResult exeuction) {
```
Done

martin-fleck-at commented on bundles/org.eclipse.emfcloud.modelserver.emf/src/org/eclipse/emfcloud/modelserver/emf/common/DefaultSessionController.java:
```diff
@@ -124,13 +125,13 @@ public void modelUpdated(final String modeluri) {
          return;
       }
-      broadcastIncrementalUpdates(modeluri, encodings);
+      broadcastIncrementalUpdates(modeluri, exeuction);
```
Done

martin-fleck-at commented:
@ndoschek I pushed an update with all the requested changes. I also moved the code a little bit around and renamed CommandCodecContribution to CommandContribution to avoid any codec confusion ;-)

ndoschek reviewed (approved):
Thank you very much Martin for updating the PR, everything works now as expected! :+1: 
My previous confusion about the AddCommand is resolved as well - I tested it again with pure EMF commands and it works great as well!
Thank you again for this great improvement!
